### DISCUSSION TITLE
DB access - no route to host

### DISCUSSION THREAD

**Discussion by patrickomatik** [gravitational/teleport#20672] [Q&A] [answered]
We are hoping to deploy teleport across our org, and are currently evaluating. So far, so good - ssh and app proxies work as expected. But we are having trouble with database access.

We have a mariadb server instance on debian 11, behind a NAT connection. Database roles, users and certificates appear to be set up correctly. The server can 'see' the wider internet just fine, as well as the teleport server login web page.

Attempting to connect to the db server instance from a linux workstation gives a 'no route to host' error. The same workstation has no problem with teleport ssh activities. I'm connecting with

`tsh db connect --db-user=root --db-name=mysql db`

Starting an ssh session with the teleport server, and attempting to connect to the db server instance from the teleport server itself also gives exactly the same 'no route to host' error.

Any advice please, I expect I am missing something fundamental?

2023-01-25T10:24:21Z ERRO [DB:SERVIC] Failed to handle connection. addr:nnn.nnn.nnn.nnn:443 error:[
ERROR REPORT:
Original Error: syscall.Errno no route to host
Stack Trace:
        github.com/gravitational/teleport/lib/srv/db/common/errors.go:147 github.com/gravitational/teleport/lib/srv/db/common.ConvertConnectError
        github.com/gravitational/teleport/lib/srv/db/mysql/engine.go:272 github.com/gravitational/teleport/lib/srv/db/mysql.(*Engine).connect
        github.com/gravitational/teleport/lib/srv/db/mysql/engine.go:93 github.com/gravitational/teleport/lib/srv/db/mysql.(*Engine).HandleConnection
        github.com/gravitational/teleport/lib/srv/db/server.go:814 github.com/gravitational/teleport/lib/srv/db.(*Server).handleConnection
        github.com/gravitational/teleport/lib/srv/db/server.go:724 github.com/gravitational/teleport/lib/srv/db.(*Server).HandleConnection
        github.com/gravitational/teleport/lib/reversetunnel/transport.go:275 github.com/gravitational/teleport/lib/reversetunnel.(*transport).start
        github.com/gravitational/teleport/lib/reversetunnel/agent.go:581 github.com/gravitational/teleport/lib/reversetunnel.(*agent).handleDrainChannels.func2
        runtime/asm_amd64.s:1594 runtime.goexit
User Message: no route to host] db/server.go:726

**Comment by webvictim** [comment]
What does the IP `nnn.nnn.nnn.nnn` that you've redacted correspond to? It looks to me like your workstation can't route to the Teleport proxy address to access the MySQL listener. Do you have any connectivity restrictions on the workstations themselves?

What's the output of `curl -s https://teleport.example.com/webapi/ping`? Feel free to redact the domain name - we're looking for `mysql_listen_addr` and `mysql_public_addr` if they're set.

**Reply by patrickomatik** [reply]
Thanks for taking the time to reply, most appreciated!

`curl -s https://teleport.my_domain.com/webapi/ping`

gives

`{"auth":{"type":"local","second_factor":"otp","preferred_local_mfa":"otp","local":{"name":""},"private_key_policy":"none","has_motd":false},"proxy":{"kube":{"enabled":true,"listen_addr":"0.0.0.0:443"},"ssh":{"listen_addr":"0.0.0.0:443","tunnel_listen_addr":"0.0.0.0:443","web_listen_addr":"0.0.0.0:443","public_addr":"teleport.my_domain.com:443"},"db":{"postgres_listen_addr":"0.0.0.0:443","mysql_listen_addr":"0.0.0.0:443"},"tls_routing_enabled":true},"server_version":"11.2.1","min_client_version":"10.0.0","cluster_name":"teleport.my_domain.com"}`

`nnn.nnn.nnn.nnn` corresponds to the public IP address of the teleport server, `teleport.my_domain.com`.

The workstation is behind NAT (different subnet entirely to that used by the teleport server), but is otherwise unrestricted. Note that attempting to connect from an ssh session running on the teleport server itself also fails in the same way, so I think that eliminates any issue with connectivity between the client session (as in the machine being used by human running `tsh db connect db`) and the teleport server.

**Reply by webvictim** [reply]
What's under the `db_service` configuration block in your `/etc/teleport.yaml` file?

Specifically, what's the IP/hostname of the database that you have configured? Does `nc -vz <ip/hostname> <port>` work from the Teleport server where the `db_service` is running?

**Comment by patrickomatik** [comment, marked as answer]
Edit: see @webvictim's comment below for the solution.

Hello again!
`db.mydomain.com` has teleport.yaml of:

```version: v3
teleport:
  nodename: db
  data_dir: /var/lib/teleport
  proxy_server: teleport.mydomain.com:443
  auth_token: /tmp/token
db_service:
  enabled: "yes"
 # Matchers for database resources created with "tctl create" command.
 # For more information: https://goteleport.com/docs/database-access/guides/dynamic-registration/
  resources:
   - labels:
      "*": "*"
  # Lists statically registered databases proxied by this agent.
  databases:
   - name: db
    protocol: mysql
    uri: db.teleport.mydomain.com:3306
    static_labels:
      "env": "dev"
    dynamic_labels:
auth_service:
  enabled: "no"
ssh_service:
  enabled: "no"
proxy_service:
  enabled: "no"
```

There is no reference to `db_service` on `teleport.mydomain.com` - which is correct, I think?

`nc -vz teleport.mydomain.com 443` from `db.mydomain.com` gives `fw.mydomain.com [nn.nn.nn.nn] 443 https open` 

- where `fw.mydomain.com` is the fqdn for the public ip aliased to `teleport.mydomain.com`

`nc -vz db.teleport.mydomain.com 3306` from `db.mydomain.com` gives `DNS fwd/rev mismatch: db.teleport.mydomain.com != fw.mydomain.com
db.teleport.mydomain.com [nn.nn.nn.nn] 3306 (mysql) : No route to host`

 - which is to be expected, and is not indicative of a problem ( `teleport.mydomain.com` does not listen on 3306 and I do not have a route through the firewall that is in front of `teleport.mydomain.com` for port 3306. I only route port 443 incoming connections to `teleport.mydomain.com`.

( `nn.nn.nn.nn` is my public ip for `teleport.mydomain.com` in all cases.

The teleport server has https certificates for `teleport.mydomain.com` and `*.teleport.mydomain.com`. )

Thanks again for any suggestions! It would be great to get to the bottom of this!

**Reply by webvictim** [reply]
I think you might be misunderstanding what Teleport needs here. The Teleport DB agent process (i.e. `db_service`) needs to be able to make two connections:
- one outbound connection from itself back to `teleport.mydomain.com` on port 443 to connect a reverse tunnel back to the Teleport cluster (this is working, as per `443 https open`)
- another outbound connection from itself directly to the MySQL server that this Teleport DB agent is going to be proxying access to - this would typically be on port 3306 (and therefore needs to be open if it's blocked by a firewall)

Is `db.mydomain.com` the same machine where the MySQL database is running? If so, you probably want to configure your `uri` to `localhost:3306` instead and restart the Teleport `db_service`.

If the target MySQL database is running on a different host/port and this machine is a jumphost or similar, you'd want to get the firewall opened from this machine -> MySQL.

I hope that makes a bit more sense!

**Reply by patrickomatik** [reply]
Yes, the db is local to the host `db.mydomain.com`. Changing the URI to `localhost:3306` has resolved the `no route to host` error; now I have a different issue, presumably db cert/db config related which we will dig into (`ERROR 1105 (HY000): the MySQL Server does not support TLS required by the client`). Many thanks for your help in resolving the routing problem, which seems obvious in retrospect - very much appreciated.

**Reply by patrickomatik** [reply]
The TLS issues are resolved in this post: https://github.com/gravitational/teleport/discussions/6099
The correct foo for certificate generation is: `tctl auth sign --format=db --host=localhost --out=server --ttl=8760h` -- nb `localhost`.

**Reply by webvictim** [reply]
Great - glad you got it working!
